### Justin Ehly
#### HW1 - NLP

In [65]:
# 2. create lexical diversity scoring routine
import nltk
def lexical_diversity(text):
    voc_size = len(text)
    lex_div = len(set(text)) / voc_size
    return lex_div, voc_size

def word_percentage(word, text):
    raw_perc = text.count(word)/ len(text) * 100
    perc = "{0:.2%}".format(raw_perc/100)
    return raw_perc, perc

In [68]:


# get data from the gutenberg.org website as utf-8 text encoded text files, remove the \r\n and blank lines
import urllib

def get_reader(url):
    file = urllib.request.urlopen(url)

    with file as f:
        lines = f.readlines()

    lines = [x.decode('utf-8') for x in lines]
    lines = [x[:-4] for x in lines if x[:-4] != '']
    return lines

first = get_reader('https://www.gutenberg.org/cache/epub/14640/pg14640.txt')
second = get_reader('https://www.gutenberg.org/cache/epub/14668/pg14668.txt')
third = get_reader('https://www.gutenberg.org/cache/epub/14766/pg14766.txt')
fourth = get_reader('https://www.gutenberg.org/cache/epub/14880/pg14880.txt')
fifth = get_reader('https://www.gutenberg.org/cache/epub/15040/pg15040.txt')
sixth = get_reader('https://www.gutenberg.org/cache/epub/16751/pg16751.txt')
seventh = get_reader('https://www.gutenberg.org/cache/epub/19721/pg19721.txt')




In [67]:

readers = {'first': first,
           'second': second,
           'third': third,
           'fourth': fourth, 
           'fifth': fifth, 
           'sixth': sixth, 
           'seventh': seventh}

scores = {}

for grade, reader in readers.items():
    scores[grade] = lexical_diversity(reader)


for k,v in scores.items():
    print('%s lexical diversity score: %s, vocabulary size: %s' % (k, v[0], v[1]))


first lexical diversity score: 0.9634630659253376, vocabulary size: 1259
second lexical diversity score: 0.9634146341463414, vocabulary size: 2460
third lexical diversity score: 0.9619047619047619, vocabulary size: 3780
fourth lexical diversity score: 0.9911264851857422, vocabulary size: 6649
fifth lexical diversity score: 0.9872780784202032, vocabulary size: 10533
sixth lexical diversity score: 0.9853685597392691, vocabulary size: 14421
seventh lexical diversity score: 0.9919767325243205, vocabulary size: 9971


<font color='blue'>3. Go to http://www.gutenberg.org/ebooks/bookshelf/215  and select texts of different grade levels (e.g., fourth reader, fifth reader et al).
   Report the lexical diversity score of each. Explain whether the result was surprising.</font>

  - The initial results are surprising because I expected there to be an increase with the lexical diversity score as the grade level advances and the human brain matures.


<font color='blue'>4. Also compare the vocabulary size of the same three texts. Explain whether the result was surprising.  </font>

  - On further thought, this does make sense because as the brain is advancing, it is actually learning more faster so there are more diverse words with each longer book.
  - The 7th grade book is a story book, so while it has less vocabulary words, it may have more complex words.

<font color='blue'>5. Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself</font>

  - When comparing different skill levels of text or langauge, it is important to take into considerion both the lexical diversity score, that is the ratio of the unique words or tokens to the total vocabulary size well as the vocabulary size itself. Simply comparing lexical diversity scores between different grade level reading texts there is no substantial observable difference with a range from .96 to .99 from first through seventh grade. When vocabulty size of those among the different texts is compared, it becomes apparent that there is a wider range of the number of words used starting with 1,259 words in the first grade version and 14,421 words in the sixth grade version. It should be noted that the seventh grade version was a compilation of stories versus a more straight forward syntactical guide found in the other texts. By utilizing the vocabulary size in conjunction with the lexical diversity score, it can be concluded that the level of complexity remains proportionate the size of the vocabulary as the texts progress numerically. In other words, the texts include more diverse vocabulary at a higher rate of speed.